src/content/en/fundamentals/media/manipulating/live-effects.md

```diff
@@ -0,0 +1,376 @@
+- Send your source image to the GPU as a texture (image)
+- Create a 'vertex shader'
+- Create a 'fragment shader'
+- Set some shader variables, called 'uniforms'
+- Run the shaders
+
+You allocate some memory on the graphics card called a vertex buffer, and you store data in it that
+describes each point of each triangle. You can also set some variables, called uniforms, that are
+global values through both shaders.
+
+A vertex shader uses data from the vertex buffer to calculate where on the screen to draw the three
+points of a triangle.
+
+Now the GPU knows which pixels within the canvas need to be drawn. The fragment shader is called
+once per pixel, and needs to return the color that should be drawn to the screen. The fragment
+shader can read information from one or more textures to determine the color.
+
+When reading a texture in a fragment shader you specify which part of the image you want to read
+using two floating-point coordinates between 0 (left or bottom) and 1 (right or top).
+
+If you want to read the texture based on pixel coordinates then you need to pass the size of the
+texture in pixels as a uniform vector so that you can do the conversion for each pixel.
+
+    // ...
+
+    varying vec2 pixelCoords;
+
+    uniform vec2 textureSize;
+    uniform sampler2D textureSampler;
+
+    main() {
+      vec2 textureCoords = pixelCoords / textureSize;
+
+      gl_FragColor = texture2D(textureSampler, textureCoords);
+
+      // ...
+    }
+
+Pretty much every kind of 2D image manipulation that you might want to do can be done in the
+fragment shader, and all of the other WebGL parts can be abstracted away. You can see
+[the abstraction layer](https://github.com/GoogleChrome/snapshot/blob/master/src/image-shader.ts)
+(in TypeScript) that is being in used in one of our sample applications if you'd like to see the
+details.
+
+### Which should I use?
+For pretty much any professional quality image manipulation, you should use WebGL. There is no
+getting away from the fact that this kind of work is the whole reason GPUs were invented. You can
+process images much, much faster on the GPU, which is essential for any real-time effects.
+
+2D canvas is much simpler, so is great for prototyping and may be fine for one-off transformations.
+However, there are plenty of abstractions around for WebGL that mean you can get the performance
+boost without needing to learn the details.
+
+If even if you parallelize your code with `Worker`s, your GPU may have 100s of times as many
+specialized cores as your CPU has general cores.
+
+Examples in this article are mostly for 2D canvas, but the principles should translate pretty easily
+to fragment shader code.
+
+## Effect Types
+
+### Pixel Effects
+This is the simplest category to both understand and implement. All of these transformations take
+the color value of a single pixel and passes it into a function that will give you back another
+color value to use as the output.
```

Comment:
Change 'that will give you back another color value to use as the output' to 'that returns another color value'.